### PR TITLE
component: allocate comp_dev on runtime_shared zone

### PR DESCRIPTION
**Pull request by keyonjie**
Allocate comp_dev on runtime_shared zone to use uncached address to fix
cache coherency issue, this should not impact performance too much as it
is not frequently used buffer.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>

### CODE REVIEW

**Comment by ranj063** [comment]
@keyonjie I understand we need this but lets not merge this one as is. We need to add coherent cached accesses for comp_dev along with this change

**Comment by keyonjie** [comment]
> @keyonjie I understand we need this but lets not merge this one as is. We need to add coherent cached accesses for comp_dev along with this change

As this is not a continuous bytes access so I am not quite sure using the coherence API with lock->invalidate->write->writeback->unlock for each single read/write access to every member of comp_dev could actually improve anything, comparing with using uncache address directly.

@ranj063 do you have some proposed change about this already?

**Comment by lgirdwood** [comment]
> > @keyonjie I understand we need this but lets not merge this one as is. We need to add coherent cached accesses for comp_dev along with this change
> 
> As this is not a continuous bytes access so I am not quite sure using the coherence API with lock->invalidate->write->writeback->unlock for each single read/write access to every member of comp_dev could actually improve anything, comparing with using uncache address directly.
> 
> @ranj063 do you have some proposed change about this already?

We should not be acquiring/release on a word at a time basis, but acquire and then do multiple word IO before release.  We need to identify this in our code to minimize locking and cache ops, but using uncache all the time is actually worse.

**Comment by ranj063** [comment]
> As this is not a continuous bytes access so I am not quite sure using the coherence API with lock->invalidate->write->writeback->unlock for each single read/write access to every member of comp_dev could actually improve anything, comparing with using uncache address directly.

@keyonjie we do access comp_dev on every copy more than once for every component (to get the component specific data pointer) in the pipeline isnt it? Im surprised you say that this will not lead to any performance increase.

**Review by ranj063 — APPROVED** [review]

**Comment by keyonjie** [comment]
> > As this is not a continuous bytes access so I am not quite sure using the coherence API with lock->invalidate->write->writeback->unlock for each single read/write access to every member of comp_dev could actually improve anything, comparing with using uncache address directly.
> 
> @keyonjie we do access comp_dev on every copy more than once for every component (to get the component specific data pointer) in the pipeline isnt it? Im surprised you say that this will not lead to any performance increase.

@ranj063 I didn't mean comparing with today's cached version, just meant that I don't foresee significant improvement if we change to use coherence APIs here.

Yes, the accurate performance impact is waiting for measurement. What I can say is that I don't hit any performance issue with the PR change yet, so it looks fine to me to proceed multi-core support with it for v2.0. 

> 
> But nevertheless @lgirdwood this will allow us to pass most of our multi-core tests. So should we go ahead and merge this for SOF2.0 and I'll try my best to complete the coherency part if I can.

**Comment by lyakh** [comment]
I'd really like to understand what are the cases of cross-core access to these `struct comp_dev` objects. In theory as long as we don't have component-level secondary core offloading, all those objects should only be accessed by the core, running the pipeline. Can we also add a link to the PR to an issue or to issues, that this is fixing? If none have been created, maybe we need to create some or at least add a description here.

**Comment by lgirdwood** [comment]
@kkarask @wszypelt looks like CI may have got stuck, can you check it. Thanks !

**Comment by lgirdwood** [comment]
I'm good to merge after this passes CI as the base starting point for stability. We can then build on this for performance as using uncache is slow.

**Comment by keyonjie** [comment]
> @kkarask @wszypelt looks like CI may have got stuck, can you check it. Thanks !

yes, looks the ICL fails for every PR, maybe DUT reboot needed?

**Comment by keyonjie** [comment]
> I'm good to merge after this passes CI as the base starting point for stability. We can then build on this for performance as using uncache is slow.

the UP2 nocodec timeout doesn't looks related, but let's re-trigger it anyway.

**Comment by keyonjie** [comment]
> I'd really like to understand what are the cases of cross-core access to these `struct comp_dev` objects. In theory as long as we don't have component-level secondary core offloading, all those objects should only be accessed by the core, running the pipeline. Can we also add a link to the PR to an issue or to issues, that this is fixing? If none have been created, maybe we need to create some or at least add a description here.

@lyakh share the same hope to address the exact point which lead to failures, but it's not an easy task to find that in short time.
As the topology change is not merged to main yet so CI can't capture these failures and reporting issues on a personal branch doesn't sound good, let me add some comments to the code change.

**Comment by keyonjie** [comment]
SOFCI TEST

**Review by lgirdwood — APPROVED** [review]

**Pull request merged**
